src/mame/drivers/generalplus_gpl_unknown.cpp

```diff
@@ -356,37 +313,174 @@ uint16_t pcp8718_state::unk_7abf_r()
 
 uint16_t pcp8718_state::unk_7860_r()
 {
-	LOGMASKED(LOG_GPL_UNKNOWN,"%06x: unk_7860_r (IO port)\n", machine().describe_context());
-
-	uint16_t ret = (m_io_p2->read() & 0xfff7);
+	uint16_t ret = (m_io_in0->read() & 0xffe7);
 
-	if ((m_7860 & 0x20))
+	if (m_bit10state)
 		ret |= 0x08;
 
+	if (m_latchinbit)
+		ret |= 0x10;
+
```

Comment:
This is a situation where `IPT_CUSTOM` and either `PORT_CUSTOM_MEMBER` or `PORT_READ_LINE_MEMBER` should be used.

Reply:
maybe, although our entire concept of INPUT_PORT for that kind of thing has always felt a bit iffy to me, especially now we have people doing outputs in there too (where the ordering can be important)

they seem to have turned more into bit-based memory maps for specific accesses, when originally they were designed for user controlled inputs / development aids / to represent the control panel.  I do find that overloading of them to be a little uncomfortable, which is why I tend to just do it in code.

Reply:
anyway I'll push my latest where I'd already started making such changes.